File: web/lib/ports.js
'use strict';

/**
 * Format → port resolution for WARP relays (JS mirror of core/ports.py).
 *
 * Ports are stored as [wireguard_port, amneziawg_port].
 * The resolver uses a strict positional index — no preference lists needed.
 */

const FORMATS = new Set(['wireguard', 'amnezia', 'wiresock', 'clash']);

// Formats that use the AmneziaWG port (ports[1]).
const _AWG_FORMATS = new Set(['amnezia', 'wiresock', 'clash']);

/**
 * Return "host:port" using strict positional index selection.
 *
 * - wireguard: always uses ports[0] (WireGuard port).
 * - amnezia, wiresock, clash: use ports[1] (AmneziaWG port) when available,
 *   falling back to ports[0] if the relay only has one port.
 * - Unknown format IDs behave like wireguard (use ports[0]).
 *
 * @param {string} fmt - canonical format ID
 * @param {{ host: string, ports: number[] }} relay - relay object
 * @returns {string}
 */
function resolveEndpoint(fmt, relay) {
  if (_AWG_FORMATS.has(fmt) && relay.ports.length >= 2) {
    return `${relay.host}:${relay.ports[1]}`;
  }
  return `${relay.host}:${relay.ports[0]}`;
}

module.exports = { FORMATS, resolveEndpoint };

File: core/ports.py
"""Format → port resolution for WARP relays.

The canonical set of supported format IDs and each format's ordered list of
preferred UDP ports live here.  The resolver finds the first preferred port
that a given relay actually supports, falling back to the relay's first port
when none match.  No JSON config involved.
"""

from __future__ import annotations

FORMATS: frozenset[str] = frozenset({"wireguard", "amnezia", "wiresock", "clash"})

# Each format lists its preferred ports from most-wanted to least-wanted.
# The resolver picks the first port in this list that the relay supports.
FORMAT_PORT_PREFERENCES: dict[str, list[int]] = {
    "wireguard": [2408, 4500, 500, 1701],
    "amnezia":   [4500, 2408, 500, 1701],
    "wiresock":  [2408, 4500, 500, 1701],
    "clash":     [2408, 4500, 500, 1701],
}


def resolve_endpoint(fmt: str, host: str, ports: list[int]) -> str:
    """Return ``host:port`` using format-aware port selection.

    Picks the first port in *fmt*'s preference list that appears in *ports*.
    Falls back to ``ports[0]`` when the format is unknown or none of its
    preferred ports are supported.
    """
    preferences = FORMAT_PORT_PREFERENCES.get(fmt, FORMAT_PORT_PREFERENCES["wireguard"])
    for port in preferences:
        if port in ports:
            return f"{host}:{port}"
    return f"{host}:{ports[0]}"
